crates/rules.bzl

```diff
@@ -0,0 +1,195 @@
+# under the License.
+#
+
+load("@rules_rust//rust:rust_common.bzl", "CrateInfo")
+
+CrateDeploymentInfo = provider(
+    fields = {
+        "crate": "Crate file to deploy",
+        "metadata": "File containing package metadata",
+    },
+)
+
+def _generate_version_file(ctx):
+    version_file = ctx.file.version_file
+    if not ctx.attr.version_file:
+        version_file = ctx.actions.declare_file(ctx.attr.name + "__do_not_reference.version")
+        version = ctx.var.get("version", "0.0.0")
+
+        ctx.actions.run_shell(
+            inputs = [],
+            outputs = [version_file],
+            command = "echo -n {} > {}".format(version, version_file.path),
+        )
+    return version_file
+
+def _assemble_crate_impl(ctx):
+    print(ctx.attr.target[CrateInfo])
+    version_file = _generate_version_file(ctx)
+    args = [
+        "--srcs", ";".join([x.path for x in ctx.attr.target[CrateInfo].srcs.to_list()]),
+        "--output-crate", ctx.outputs.crate_package.path,
+        "--output-metadata-json", ctx.outputs.metadata_json.path,
+        "--root", ctx.attr.target[CrateInfo].root.path,
+        "--edition", ctx.attr.target[CrateInfo].edition,
+        "--name", ctx.attr.target[CrateInfo].name,
+        "--version-file", version_file.path,
+        "--authors", ";".join(ctx.attr.authors),
+        "--keywords", ";".join(ctx.attr.keywords),
+        "--categories", ";".join(ctx.attr.categories),
+        "--description", ctx.attr.description,
+        "--documentation", ctx.attr.documentation,
+        "--homepage", ctx.attr.homepage,
+        "--license", ctx.attr.license,
+        "--repository", ctx.attr.repository,
+        "--deps", ";".join(["{}={}".format(k, v) for k, v in ctx.attr.deps.items()]),
+    ]
+    inputs = [version_file]
+    if ctx.file.readme_file:
+        args.append("--readme-file")
+        args.append(ctx.file.readme_file.path)
+        inputs.append(ctx.file.readme_file)
+    ctx.actions.run(
+        inputs = inputs + ctx.attr.target[CrateInfo].srcs.to_list(),
+        outputs = [ctx.outputs.crate_package, ctx.outputs.metadata_json],
+        executable = ctx.executable._crate_assembler_tool,
+        arguments = args,
+    )
+    return [
+        CrateDeploymentInfo(
+            crate = ctx.outputs.crate_package,
+            metadata = ctx.outputs.metadata_json,
+        ),
+    ]
+
+assemble_crate = rule(
+    implementation = _assemble_crate_impl,
+    attrs = {
+        "target": attr.label(
+            mandatory = True,
+            doc = "`rust_library` label to be included in the package",
+        ),
+        "version_file": attr.label(
+            allow_single_file = True,
+            doc = """
+            File containing version string.
+            Alternatively, pass --define version=VERSION to Bazel invocation.
+            Not specifying version at all defaults to '0.0.0'
+            """,
+        ),
+        "deps": attr.string_dict(
```

Comment:
Why can't we infer the dependencies from crate the way we did in `assemble_maven`, @vmax ?

Reply:
@lolski I've created a GH issue so we don't forget to fix it: #323